design/one-pager-cross-resource-referencing.md

```diff
@@ -0,0 +1,296 @@
+metadata:
+  name: my-network
+  namespace: cool-ns
+spec:
+  ...
+---
+apiVersion: compute.gcp.crossplane.io/v1alpha2
+kind: Subnetwork
+metadata:
+  name: my-subnetwork
+  namespace: cool-ns
+spec:
+  ...
+  network: /projects/myproject/global/networks/example
+  ...
+```
+
+To add `networkRef` field to the `Subnetwork`, we can implement `NetworkRefResolver` as:
+
+```go
+type NetworkRefResolver struct {
+  NetowrkName string `json:"vpcName"`
+  ProjectName string `json:"projectName"`
+}
+
+func (r *NetworkRefResolver) Resolve()(string,error){
+  
+  // GetObject is a method that reutrns an API object given its name  
+  n,err := GetObject(r.NetworkName)
+  if err!= nil{
+    return error
+  }
+
+  p,err := GetObject(r.ProjectName)
+  if err!= nil{
+    return error
+  }
+
+  return fmt.Sprintf("/projects/%v/global/networks/%v", n.Name, p.Name), nil
+}
+```
+
+And then update [Subnetwork type] as:
+
+```diff
+- Network string `json:"network"`
++ NetworkRef NetworkRefResolver `json:"networkRef"`
+```
+
+Finally, the YAML object for this solution can be re-written as:
+
+```yaml
+apiVersion: compute.gcp.crossplane.io/v1alpha2
+kind: Subnetwork
+metadata:
+  name: my-subnetwork
+  namespace: cool-ns
+spec:
+  ...
+  networkRef:
+    networkName: my-network
+    projectName: myproject
+  ...
+```
+
+There are a number of simplifying assumptions in this implementation. Using the
+notation we used for resources earlier:
+
+* The API group and version of `r2` and `r1` are the same
+* The namespace of `r2` and `r1` are the same
+
+### Maintain High Fidelity
+
+When provisioning resources, it is desirable to support existing external
+resources which are not managed by Crossplane. For instance assume that in the
+`VPC` and `Subnet` sample solution, the `VPC` resource already exists and we only want
+to provision the `Subnet`. Since we changed the `VPCID` to `VPCIDRef` in the
+`Subnet` type, it won't be possible to use the external VPCID.
+
+To support this case, we need to keep the `VPCID` field in `Subnet`, so we
+change the update the modification as:
+
+```diff
+  // VPCID is the ID of the VPC.
+- VPCID string `json:"vpcId"`
++ VPCID string `json:"vpcId,omitempty"`
++ // VPCIDRef builds the VPCID from refenreced objects
++ VPCIDRef VpcIDRef `json:"vpcIdRef,omitempty"`
+```
+
+Note that we added `omitempty` rule to both fields, making them optional.
+
+This scheme has the following restrictions:
+
+* If none of `vpcId` or `vpcIDRef` are provided, result in an error
+* If both `vpcId` and `vpcIDRef` are provided, result in an error
+
+### Project Blocking Dependency in Resource Status
+
+Resolving an attribute  might return an error for different reasons:
+
+1. One or more of the referenced objects *doesn't exist*. In this case the
+   status of the resource should show an error, with a message for more details.
+
+2. One or more of the referenced objects is not yet *Ready*. In this case the
+   status of the resource should be updated to `Queued`.
```

Comment:
Would we determine readiness by inspecting the `Ready` status condition of the resource we depend on? In practice every current managed resource maintains a `Ready` condition, but in theory doing so is optional.

We used to expose an API for checking whether a particular condition type was true, but I [removed it](https://github.com/crossplaneio/crossplane/pull/499) with the intent that status conditions would be purely informational output for humans rather than something used as programmatic flow control. On reflection I'm less confident it was that bad to make programmatic decisions based on the most recent recorded status conditions of an object. This [lengthy section of the Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) may provide some guidance on whether it's appropriate to use conditions for this.

Reply:
@negz I think `Ready` makes the most sense. Are you proposing that we use a status other than Ready when resolving blocking dependencies?

Reply:
> Are you proposing that we use a status other than Ready when resolving blocking dependencies?

No - my intention was to double check that we're confident that making programmatic decisions based on _any_ status condition (i.e things in the `ConditionedStatus` array) is in compliance with the current stance of the Kubernetes API conventions on status conditions rather than suggesting `Ready` might be the wrong condition to test.

I wonder what the UX would be like if we didn't explicitly test the readiness of the resource we depend on at all, but rather proceeded to attempt to create our external resource as soon as we had enough information to do so. That is, when:

* The managed resource we depend on exists in Kubernetes. This does not necessarily mean the external resource it represents is ready, or exists at all.
* The resource spec and/or status fields of the resource we depended on are non-zero.

Reply:
I see your point. I think having an abstraction here is useful; we wait until the resource is "Ready" to be used, then we can consume it as a dependent. However, I raised  https://github.com/kubernetes/community/issues/4154 to ask for more opinion about this, as the current documentation you linked isn't very specific.